src/details/ArborX_DetailsBruteForceImpl.hpp

```diff
@@ -53,26 +53,94 @@ struct BruteForceImpl
+    int const primitives_per_team = max_scratch_size / sizeof(PrimitiveType);
+
+    int const n_primitive_tiles =
+        ceil((float)n_primitives / primitives_per_team);
+    int const n_predicate_tiles =
+        ceil((float)n_predicates / predicates_per_team);
+    int const n_teams = n_primitive_tiles * n_predicate_tiles;
+
+    using ScratchPredicateType =
+        Kokkos::View<PredicateType *,
+                     typename ExecutionSpace::scratch_memory_space,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    using ScratchPrimitiveType =
+        Kokkos::View<PrimitiveType *,
+                     typename ExecutionSpace::scratch_memory_space,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    int scratch_size = ScratchPredicateType::shmem_size(predicates_per_team) +
+                       ScratchPrimitiveType::shmem_size(primitives_per_team);
 
     Kokkos::parallel_for(
         "ArborX::BruteForce::query::spatial::"
         "check_all_predicates_against_all_primitives",
-        Kokkos::MDRangePolicy<ExecutionSpace, Kokkos::Rank<2>>(
-            space, {0, 0}, {n_primitives, n_predicates}),
-        KOKKOS_LAMBDA(int primitive_index, int predicate_index) {
-          auto const &predicate =
-              AccessPredicates::get(predicates, predicate_index);
-          auto const &primitive =
-              AccessPrimitives::get(primitives, primitive_index);
-          if (predicate(primitive))
+        TeamPolicy((long)n_teams, Kokkos::AUTO)
+            .set_scratch_size(0, Kokkos::PerTeam(scratch_size)),
+        KOKKOS_LAMBDA(const typename TeamPolicy::member_type &teamMember) {
+          // select the tiles of predicates/primitives checked by each team
+          int predicate_start = predicates_per_team *
+                                (teamMember.league_rank() / n_primitive_tiles);
+          int primitive_start = primitives_per_team *
+                                (teamMember.league_rank() % n_primitive_tiles);
+
+          int predicates_in_this_team = KokkosExt::min(
+              predicates_per_team, n_predicates - predicate_start);
+          int primitives_in_this_team = KokkosExt::min(
+              primitives_per_team, n_primitives - primitive_start);
+
+          ScratchPredicateType scratch_predicates(teamMember.team_scratch(0),
+                                                  predicates_per_team);
+          ScratchPrimitiveType scratch_primitives(teamMember.team_scratch(0),
+                                                  primitives_per_team);
+          // rank 0 in each team fills the scratch space with the
+          // predicates / primitives in the tile
+          if (teamMember.team_rank() == 0)
           {
-            callback(predicate, primitive_index);
+            Kokkos::parallel_for(
+                Kokkos::ThreadVectorRange(teamMember,
+                                          (long)predicates_in_this_team),
```

Comment:
Again, is there a reason for the cast here? The original variable is of type `int`, the type for the maximum index is `int` in the line below.